comment/src/comment.ts

```diff
@@ -0,0 +1,280 @@
+  return data === undefined || data.lineComment === undefined
+    ? null
+    : new LineCommenter(data.lineComment).toggle(option, state)
+}
+
+// This class performs toggle, comment and uncomment
+// of block comments in languages that support them.
+// The `open` and `close` arguments refer to the open and close
+// tokens of which this `BlockCommenter` is made up.
+export class BlockCommenter {
+  open: string
+  close: string
+  margin: string
+  constructor(open: string, close: string, margin: string = " ") {
+    this.open = open
+    this.close = close
+    this.margin = margin
+  }
+
+  toggle(option: CommentOption, state: EditorState): Transaction | null {
+    const selectionCommented = this.isSelectionCommented(state)
+    if (selectionCommented !== null) {
+      if (option !== CommentOption.OnlyComment) {
+        const tr = state.t()
+        const mapRef = tr.mapRef()
+        for (const range of selectionCommented) {
+          const open = range.open
+          const close = range.close
+          open.pos = mapRef.mapPos(open.pos)
+          tr.replace(open.pos - this.open.length, open.pos + open.margin, "")
+          close.pos = mapRef.mapPos(close.pos)
+          tr.replace(close.pos - close.margin, close.pos + this.close.length, "")
+        }
+
+        return tr
+      }
+    } else {
+      if (option !== CommentOption.OnlyUncomment) {
+        const tr = state.t()
+        tr.forEachRange((range: SelectionRange, tr: Transaction) => {
+          const copen = new Change(range.from, range.from, tr.startState.splitLines(this.open + this.margin))
+          const cclose = new Change(range.to, range.to, tr.startState.splitLines(this.margin + this.close))
+          tr.change([copen, cclose])
+          const shift = (this.open + this.margin).length
+          return new SelectionRange(range.anchor + shift, range.head + shift)
+        })
+
+        return tr
+      }
+    }
+
+    return null
+  }
+
+  // Determines whether all selection ranges in `state` are block-commented.
+  isSelectionCommented(state: EditorState): { open: { pos: number, margin: number }, close: { pos: number, margin: number } }[] | null {
+    let result = []
+    for (const range of state.selection.ranges) {
+      const x = this.isRangeCommented(state, range)
+      if (x === null) return null
+      result.push(x)
+    }
+    return result
+  }
+
+  // Determines if the `range` is block-commented in the given `state`.
+  // The `range` must be a valid range in `state`.
+  isRangeCommented(state: EditorState, range: SelectionRange): { open: { pos: number, margin: number }, close: { pos: number, margin: number } } | null {
+    type SearchWithType = (this: string, searchString: string, pos?: number) => boolean
+    const search = (pos: number, searchString: string, searchWith: SearchWithType, d: 1 | -1, i: 1 | 0): { pos: number, margin: number } | null => {
+      const line = state.doc.lineAt(pos)
+      const str = line.content as string
```

Comment:
`line.content` isn't public (and not guaranteed to hold a string). It might make more sense to create a text cursor instead of accessing by-line?

Reply:
Ok, I'll look into that.

Reply:
If I understand this properly, you mean to use `Text.iterLines` to get a `TextIterator` (which is a `LineCursor`) and go from there? If that's the case, I'll have to search twice in `getLinesInRange`, one to get the line using `doc.lineAt` and the one mentioned above. Maybe there is another way. Would it be possible to get the content from a `Line`? Sure it would be handy, not sure about secondary effects.

Reply:
You can also use [`Line.slice`](https://codemirror.net/6/docs/ref/#text.Line.slice)

Reply:
Exactly what I needed. Thanks!